### DISCUSSION TITLE
ZK deterministic address

### DISCUSSION THREAD

**Discussion by johnpaulcas** [zkSync-Community-Hub/zksync-developers#30] [Deployments] [answered]
I tried to deploy contracts on mainnet eth and zksync era, Im using same address and nonce but my deployment contracts end up different contract address? but when I deployed to eth and polygon and bsc it generate same contract address

is there any difference on zkSync era why its not generating same deterministic address?

**Comment by dutterbutter** [comment, marked as answer]
Hey @johnpaulcas 👋 Yes there are differences. You can review [this part of the docs ](https://era.zksync.io/docs/reference/architecture/contracts/contract-deployment.html#differences-in-create-behaviour) that describes how zkSync employs a unique account model and method for generating contract addresses. Instead of just relying on the sender's address and nonce like Ethereum, BSC, or Polygon, zkSync splits the nonce into two parts: the deployment nonce and the transaction nonce. This is in order to facilitate [support for account abstraction](https://era.zksync.io/docs/reference/concepts/aa).

To have a deterministic address, you should use the `create2` method from [ContractDeployer](https://era.zksync.io/docs/reference/architecture/contracts/system-contracts.html#contractdeployer). It is available for EOAs as well.

**Reply by ZhimaoL** [reply]
Hi @dutterbutter, it makes sense to use create2 from the ContractDeployer to get a deterministic address, some quick follow-up questions:

1. I use a deployer contract with the create2 method on Ethereum, Op, and Arb to get the same deployed addresses. Even though the contract itself is the same, I cannot get the same deployed address again on Era because there is no "customized deployer contract" on Era and we have to use a system deployer contract, right? 

2. Anyway to compute the counterfactual address? Similar to openzepplin's [computeAddress](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/cd981f6521e22f2b74ee9d1cea27c6b16d318528/contracts/utils/Create2.sol#L73).

3. All my contract factories previously reply on customized deployer contracts need to switch to call create2 on system-level ContractDeployer directly right? Not sure if this is the best practice. 


Thanks a lot!

**Reply by dutterbutter** [reply]
Hey hey 👋 , 

(1) Yes, that is correct you need to make use of the system contract, `ContractDeployer` 
(2) We have different address derivation from EVM, so the formula should be slightly modified. You can refer to this implementation here 👉 [getNewAddressCreate2](https://github.com/matter-labs/system-contracts/blob/e96dfe0b5093fa95c2fb340c0411c646327db921/contracts/ContractDeployer.sol#L91)
(3) Not exactly sure your system requirements, but again need to make of the system contract, [ContractDeployer](https://github.com/matter-labs/system-contracts/blob/main/contracts/ContractDeployer.sol)

Hope that helps! Happy hacking.

**Comment by varnerrobertv** [comment]
In zkSync, contract deployment differs from traditional EVM-based networks like Ethereum, Polygon, and Binance Smart Chain. While these networks use deterministic address generation based on the sender's address and nonce, zkSync employs a different mechanism called the incremental nonce for contract deployment. This means that even if the sender's address and nonce are the same, the resulting contract address may vary due to zkSync's unique addressing scheme. Therefore, it's expected behavior that contract addresses generated on zkSync may differ from those on Ethereum, Polygon, or Binance Smart Chain.

**Comment by mcgeesusan** [comment]
zkSync Era utilizes a different consensus mechanism and network architecture compared to Ethereum, Polygon, and BSC, which may result in variations in contract address generation. While Ethereum, Polygon, and BSC employ deterministic address generation algorithms, zkSync Era's design factors may lead to different contract addresses despite using the same address and nonce.

**Comment by matthewredmanv** [comment]
On zkSync Era, the contract deployment process utilizes a different mechanism compared to Ethereum, Polygon, and Binance Smart Chain. zkSync Era employs zk-rollup technology, which may result in different contract addresses due to the underlying consensus and validation mechanisms used in zk-rollup solutions.

**Comment by glasrudmary** [comment]
It seems that when deploying contracts on mainnet Eth and ZkSync Era with the same address and nonce, the resulting contract addresses are different. However, when deploying on other chains such as Polygon and BSC, the same contract address is generated. This may be due to differences in the deployment process or address generation algorithm on ZkSync Era compared to other chains. Further investigation may be needed to determine the exact cause of this discrepancy.
